File: 5/Shape.java
/*
 * started by: M. Moussavi
 * Date: Feb 2015
 * Modified by: Mitchell Sawatzky
 */
abstract class Shape implements Cloneable, Resizeable, Accessible
{
	protected Point origin;
	protected Text name;
	abstract protected Double area();
	abstract protected Double perimeter();
	abstract protected Double volume();

	protected Shape(Double x_origin, Double y_origin, String name, Colour colour){

		origin = new Point(x_origin,y_origin, colour);
		this.name = new Text(name);
	}

	protected Point  getOrigin()
	{
		return origin;
	}

    public Object clone() throws CloneNotSupportedException {
        Shape obj = (Shape)super.clone();
        obj.origin = (Point)origin.clone();
        obj.name = (Text)name.clone();

        return obj;
    }

	protected  Double distance(   Shape  other)
	{
		return origin.distance(other.origin);
	}

	protected Double  distance(   Shape  a,    Shape  b)
	{
		return Point.distance(a.origin, b.origin);
	}


	protected void  move(Double dx, Double dy)
	{
		origin.setx(origin.getx()+dx);
		origin.sety(origin.gety()+dy);
	}

	@Override
	public String toString(){
		String s = "\nShape name: " + name + "\nOrigin: " + origin;
		return s;
	}

	public String getName() {
		return name.getText();
	}

	public void setName(String newName) {
		name.setText(newName);
	}
}

File: 9/shape.cpp
// File: shape.cpp
// Author: Mitchell Sawatzky
// Date: March 27, 2016
// Class: ENSF 409

#include "shape.h"
#include "point.h"
#include <iostream>
using namespace std;

Shape::Shape (double x, double y, const char* name) : origin(x, y) {
    shapeName = new char[strlen(name) + 1];
    strcpy(shapeName, name);
}

Shape::Shape (const Shape& src) : origin(src.getOrigin().getX(), src.getOrigin().getY()) {
    shapeName = new char[strlen(src.getName()) + 1];
    strcpy(shapeName, src.getName());
}

Shape::~Shape () {
    delete [] shapeName;
}

Shape& Shape::operator= (const Shape& rhs) {
    if (this != &rhs) {
        if (shapeName) {
            delete [] shapeName;
        }
        origin = Point(rhs.getOrigin().getX(), rhs.getOrigin().getY());
        shapeName = new char[strlen(rhs.getName()) + 1];
        strcpy(shapeName, rhs.getName());
    }
    return *this;
}

const Point& Shape::getOrigin () const {
    return origin;
}

char* Shape::getName () const {
    return shapeName;
}

void Shape::display () const {
    cout << "  Shape Name: " << shapeName << endl;
    origin.display();
}

double Shape::distance (Shape& other) {
    return Point::distance(origin, other.getOrigin());
}

double Shape::distance (Shape& the_shape, Shape& other) {
    return Point::distance(the_shape.getOrigin(), other.getOrigin());
}

void Shape::move (double dx, double dy) {
    origin.setX(origin.getX() + dx);
    origin.setY(origin.getY() + dy);
}
